pkg/attribute/mutableBag.go

```diff
@@ -96,10 +111,58 @@ func copyValue(v interface{}) interface{} {
 	return v
 }
 
-// Done indicates the bag can be reclaimed.
-func (mb *MutableBag) Done() {
+// Child allocates a derived mutable bag.
+func (mb *MutableBag) Child() *MutableBag {
+	return GetMutableBag(mb)
+}
+
+// childDone indicates that a child bag has been reclaimed.
+func (mb *MutableBag) childDone() {
+	mb.Lock()
+	mb.children--
+	// if this bag was already scheduled for done()
+	// then perform the operation.
+	if mb.doneScheduled && mb.children == 0 {
+		mb.done()
+	}
+	mb.Unlock()
+}
+
+// BagDone releases a bag
+// calling thread should not use it afterwards.
+// Sets the underlying pointer to nil after releasing.
+// This ensures that use after free results in a hard failure.
+func BagDone(bPtr **MutableBag) {
+	if bPtr == nil {
+		return
+	}
+	mb := *bPtr
+	mb.doneInternal()
+	*bPtr = nil
+}
+
+// doneInternal indicates the bag can be reclaimed.
+func (mb *MutableBag) doneInternal() {
+	mb.Lock()
+	mb.doneScheduled = true
+	if mb.children == 0 {
+		mb.done()
+	}
+	mb.Unlock()
+}
+
+// done reclaims the bag. always called under a lock to the bag
+func (mb *MutableBag) done() {
 	mb.Reset()
```

Comment:
so Reset() doesn't clear the parent of the bag nor does it clear the `doneScheduled` field?  Is there ever a context where we want only the contents of the bag reset, but not the other metadata of the bag?